### ISSUE TITLE
Handler of all failed messages by handling IFailed<object>

### ISSUE THREAD

**Issue by jens-jonsson** [rebus-org/Rebus#805]
Trying to create a handler to handle all failed messages by implementing `IHandleMessages<IFailed<object>>`.

Documentation suggests that implementing `IHandleMessages<object>` will call the handler for all messages but my constraint of IFailed<object> does not seem to create a match when resolving handlers.

I am on .net core using default IServiceProvider. Checking assembly scanning shows that my handler gets registered but not called for failed messages.

**Comment by mookid8000** [comment]
I'm not sure if that will work 😁 

First thing to look for would be what your IoC container returns when resolving a list of `IHandleMessages<IFailed<object>>`

**Comment by jens-jonsson** [comment]
Sorry should have been more clear on that.

Calling `IServiceProvider.GetServices<IHandleMessages<IFailed<object>>>()` returns a list containing an instance of my registered handler.

**Comment by mookid8000** [comment]
> Calling `IServiceProvider.GetServices<IHandleMessages<IFailed<object>>>()` returns a list containing an instance of my registered handler.

oh sorry, what I meant was this: Try and see what the container returns when resolving a list of `IHandleMessages<IFailed<YourMessageType>>`

**Comment by jens-jonsson** [comment]
`IServiceProvicer.GetServices<IHandleMessages<IFailed<PublishEntityCommand>>>` works fine too, returning a list with my handler in it.

Base problem seems to be that stacking generics like `IMessageHandler<T>` where T is `IFailed<U>` does not work as I thought.

I guess if `IFailed<>` implemented e.g. a new marker interface `IFailed` I could use `IHandleMessages<IFailed>` and then work my way to what type the wrapped message is and what exception lies behind the failure.

Basically I am looking for a way to intercept all failed messages (after all retries have failed) regardless of type to be able to send it to telemetry for tracking. Much just like the example of using `IFailFastChecker` enables you to hook into the retry mechanism before the message fails permanently.

Decorating IErrorHandler works but context is lost and by that also our original correlationId.

**Comment by tompazourek** [comment]
I've just been experimenting with this. My trick is just to use a generic handler:

```csharp
internal class AnyFailedMessageHandler<T> : IHandleMessages<IFailed<T>>
{
    public async Task Handle(IFailed<T> message)
    {
        // ...
    }
}
```

I don't know many DI containers very well, but my intuition is that it's the DI container that might be the limitation.

In my example, I am using Simple Injector with automatic registration of handlers:

```csharp
Assembly[] assembliesWithHandlers = { /* ... */ };

var messageHandlerTypes = container.GetTypesToRegister
(
    typeof(IHandleMessages<>),
    assembliesWithHandlers,
    new TypesToRegisterOptions
    {
        IncludeComposites = false,
        IncludeDecorators = false,
        IncludeGenericTypeDefinitions = true, // this is the stuff!
    },
);

container.Collection.Register(typeof(IHandleMessages<>), messageHandlerTypes);
```
